### PR TITLE
Bridge azure properties to spring starter to authenticate

### PR DESCRIPTION
**Pull request by moarychan**
Fixes Azure/azure-sdk-for-java#21971

Based on the target PR changes, try to use each starter property class and the super `AzureProperties` to construct `TokenCredential` to authenticate.

Changelog;
- Add a new `ChainedTokenCredential` bean, then it will be used as the final token credential to each service client builder.

  - The first is each module's `TokenCredential`, it's created by `SpringMappingCredentialPropertiesProvider`.
      `MappingCredentialPropertiesProvider.java` interface to mapping each module properties to 
      `MappingCredentialPropertiesProvider` implementation, then rach module will use `SpringEnvironmentCredentialBuilder` 
      object to build a starter's `TokenCredential` which will be used as a credential option.
      Provider `DefaultQueueClientBuilderCustomizer` to cover the common use-cases, and combine the 
  - The second is the Azure Spring's common `TokenCredential`, it's created from `AzureDefaultTokenCredentialAutoConfiguration` auto-configuration.
- Use `Configurer` and `Customizer` pattern to apply each module service client builder definition. 
   The abstract class `AbstractClientBuilderConfigurer` to customize the special configured.
   
  Below interfaces are credential customizer:
    - `AzureKeyCredentialClientBuilderCustomizer`
    - `ConnectionStringClientBuilderCustomizer`
    - `SharedKeyCredentialClientBuilderCustomizer`
    - `TokenCredentialClientBuilderCustomizer`

Todolist:
- [ ] Properties validation if not configured any credential information.
- [x] Cosmos
- [x] Storage Blob
- [x] Storage Queue
- [ ] [WIP] KeyVault
- [ ] EventHubs
- [ ] Service Bus
- [ ] Update and add Unit testcases
- [ ] Test with all related samples
- [ ] Update dev docs

### CODE REVIEW

File: sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureDefaultTokenCredentialAutoConfiguration.java [hunk 1]
```diff
@@ -24,6 +28,7 @@
  * Auto-configuration for Azure Spring default token credential.
  */
 @Configuration
+@ConditionalOnProperty(name = AzureProperties.PREFIX)
```
[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureDefaultTokenCredentialAutoConfiguration.java:31]: No, I will remove it

File: sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java [hunk 2]
```diff
@@ -27,37 +39,102 @@
...
+        chainedTokenCredentialBuilder.addLast(propertiesProvider.mappingTokenCredential())
+                                     .addLast(defaultTokenCredential);
+        return chainedTokenCredentialBuilder.build();
+    }
+
     @Bean
-    public AzureKeyCredential azureKeyCredential() {
-        return new AzureKeyCredential(cosmosProperties.getKey());
+    @ConditionalOnMissingBean
+    public ClientBuilderCustomizer<CosmosClientBuilder> cosmosClientBuilderCustomizers() {
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java:78]: How about putting this into the builder creation method?

File: sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosClientBuilderConfigurer.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    private AzureKeyCredentialClientBuilderCustomizer<CosmosClientBuilder> azureKeyCredentialCustomizer;
+
+    public void setAzureKeyCredentialCustomizer(AzureKeyCredentialClientBuilderCustomizer<CosmosClientBuilder> azureKeyCredentialCustomizer) {
+        this.azureKeyCredentialCustomizer = azureKeyCredentialCustomizer;
+    }
+
+    @Override
+    public CosmosClientBuilder configure(CosmosClientBuilder builder) {
+        configureClientBuilder(builder);
+        if (azureKeyCredentialCustomizer != null) {
```
Comment (saragluna) [inline sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosClientBuilderConfigurer.java:27]: How about omitting the DefaultSkipCredentialCallback and setting the credential by precedence, first setting the lower precedence and higher precedence.

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java [hunk 4]
```diff
@@ -27,44 +40,96 @@
...
+        return properties.getDatabase();
+    }
+
+    @Bean(COSMOS_AZURE_KEY_CREDENTIAL_BEAN_NAME)
+    @ConditionalOnMissingBean(name = COSMOS_AZURE_KEY_CREDENTIAL_BEAN_NAME)
+    public AzureKeyCredential cosmosAzureKeyCredential() {
+        return Optional.ofNullable(properties.getKey())
+                       .filter(StringUtils::hasText)
+                       .map(AzureKeyCredential::new)
+                       .orElse(null);
```
Comment (stliu) [inline sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java:64]: null?

File: sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java [hunk 5]
```diff
@@ -27,44 +40,96 @@
...
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public AzureKeyCredentialClientBuilderCustomizer<CosmosClientBuilder> azureKeyCredentialCustomizer(
+        @Autowired(required = false) @Qualifier(COSMOS_AZURE_KEY_CREDENTIAL_BEAN_NAME) AzureKeyCredential cosmosAzureKeyCredential) {
+        if (cosmosAzureKeyCredential != null) {
+            return builder -> builder.credential(cosmosAzureKeyCredential);
+        }
+        return null;
```
Comment (stliu) [inline sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java:85]: we should discuss this design, I'm really not sure if it is the right way to do this

Comment (stliu) [reply sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/cosmos/CosmosAutoConfiguration.java:85]: btw, you can use `ObjectProvider` to get rid of this `@Autowired(required=false)`

**Comment by saragluna** [comment]
Already merged in this PR https://github.com/Azure/azure-sdk-for-java/pull/23787. Closing this one now.

**Pull request closed without merging**